# Inventory Management

AOC Day 5 Part 1

```
3-5
10-14
16-20
12-18

1
5
8
11
17
32
```

The fresh ID ranges are inclusive: the range 3-5 means that ingredient IDs 3, 4, and 5 are all fresh. The ranges can also overlap; an ingredient ID is fresh if it is in any range.

The Elves are trying to determine which of the available ingredient IDs are fresh. In this example, this is done as follows:

- Ingredient ID 1 is spoiled because it does not fall into any range.
- Ingredient ID 5 is fresh because it falls into range 3-5.
- Ingredient ID 8 is spoiled.
- Ingredient ID 11 is fresh because it falls into range 10-14.
- Ingredient ID 17 is fresh because it falls into range 16-20 as well as range 12-18.
- Ingredient ID 32 is spoiled.

So, in this example, 3 of the available ingredient IDs are fresh.

In [282]:
input = [
    "3-5",
    "10-14",
    "16-20",
    "12-18",
    "",
    "1",
    "5",
    "8",
    "11",
    "17",
    "32",
]

Let us process and split the inputs.

In [283]:
split_idx = input.index("")

ranges, ids = input[:split_idx], input[split_idx + 1:]
print(f"Ranges: {ranges}, IDs: {ids}")

Ranges: ['3-5', '10-14', '16-20', '12-18'], IDs: ['1', '5', '8', '11', '17', '32']


We can split and process the ranges as follows:

```python
for range in ranges:
    split = range.strip().split("-")
    l, r = split[0], split[1]
    pass
```

We will first consider the available ingredient IDS.

In [284]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.

    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [285]:
# Sort ranges by the integer value of the left endpoint, descending
ranges = sorted(ranges, key=lambda x: int(x.split("-")[0]))
ranges

['3-5', '10-14', '12-18', '16-20']

In [286]:
def bin_search(ranges: list[str], elem: int):
    left = 0
    right = len(ranges) - 1
    while left <= right:
        mid = (left + right) // 2
        mrs = ranges[mid].split("-")
        # left and right margins of the range
        l, r = int(mrs[0]), int(mrs[1])
        if elem in range(l, r+1):
            return 1
        elif l < elem:
            left = mid + 1
        elif r > elem:
            right = mid - 1
    return 0

In [287]:
ranges

['3-5', '10-14', '12-18', '16-20']

In [288]:
print(f"1 -> {bin_search(ranges, 1)}")
print(f"3 -> {bin_search(ranges, 5)}")
print(f"5 -> {bin_search(ranges, 5)}")
print(f"8 -> {bin_search(ranges, 8)}")
print(f"11 -> {bin_search(ranges, 11)}")
print(f"17 -> {bin_search(ranges, 17)}")
print(f"32 -> {bin_search(ranges, 32)}")

1 -> 0
3 -> 1
5 -> 1
8 -> 0
11 -> 1
17 -> 1
32 -> 0


Looks like we are good.

**A small note**: `17` ideally fits two ranges: `12-18` and `16-20`. Binary search returns only one. But for our case, it doesn't matter because we need to count it only once.

So, let's wrap the algorithm into a function.

In [289]:
def search(ranges: list[str], elem: int):
    ranges = sorted(ranges, key=lambda x: int(x.split("-")[0]))
    return bin_search(ranges=ranges, elem=elem)

Let us call it for the input.

In [290]:
count = 0

print(ids)
ranges, ids = input[:split_idx], input[split_idx + 1:]

ranges.remove("") if "" in ranges else {}
ids.remove("") if "" in ids else {}

for id in ids:
    count += search(ranges=ranges, elem = int(id))

print(f"Count is {count}")

['1', '5', '8', '11', '17', '32']
Count is 3


Let's wrap our entire algorithm into a function!

In [291]:
def f(input: list[str]) -> int:
    split_idx = input.index("")
    print(f"Splitting the ranges at {split_idx}")
    ranges, ids = input[:split_idx], input[split_idx + 1:]

    print(f"Ranges: {len(ranges)} and IDs: {len(ids)}")
    print(f"Ranges start from {ranges[0]}, and IDs start from {ids[0]}")
    print(f"Ranges end with {ranges[len(ranges) - 1]}, and IDs end with {ids[len(ids) - 1]}")

    count = 0
    for id in ids:
        count += search(ranges=ranges, elem=int(id))
    return count

Let's run it for our input.

In [292]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]
    print(f"Count: {f(input)}")


Splitting the ranges at 169
Ranges: 169 and IDs: 1000
Ranges start from 213205509444883-217326119178383, and IDs start from 417270233264984
Ranges end with 88583581137116-89047622016807, and IDs end with 58609737910842
Count: 606


The number $606$ seems to be incorrect. Let's troubleshoot.

In [293]:
x = 86051125252015
type(x)

int

In [294]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    split_idx = input.index("")
    ranges, ids = input[:split_idx], input[split_idx + 1:]

    count = 0
    for id in ids:
        count += search(ranges=ranges, elem=int(id))

    print(f"Count = {count}")

Count = 606


Let us revert back to a naive $O(n^2)$ solution.

In [295]:
input = [
    "3-5",
    "10-14",
    "16-20",
    "12-18",
    "",
    "1",
    "5",
    "8",
    "11",
    "17",
    "32",
]

split_idx = input.index("")

ranges, ids = input[:split_idx], input[split_idx + 1:]
count = 0

for id in ids:
    for r in ranges:
        range_split = r.split("-")
        l, r = int(range_split[0]), int(range_split[1])
        if int(id) in range(l, r+1):
            count += 1
            break   # Because of overlapping

print(f"Count = {count}")

Count = 3


In [296]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    split_idx = input.index("")

    ranges, ids = input[:split_idx], input[split_idx + 1:]
    count = 0

    for id in ids:
        for r in ranges:
            range_split = r.split("-")
            l, r = int(range_split[0]), int(range_split[1])
            if int(id) in range(l, r+1):
                count += 1
                break   # Because of overlapping

print(f"Count = {count}")

Count = 623


Well, $623$ seems to be the correct answer. There is something up with the binary search way of doing it.

If this was done via binary search, the complexity would be $O(nlog(n))$, instead of $O(n^2)$.

```python
def bin_search(ranges: list[str], elem: int):
    left = 0
    right = len(ranges) - 1
    while left <= right:
        mid = (left + right) // 2
        mrs = ranges[mid].split("-")
        # left and right margins of the range
        l, r = int(mrs[0]), int(mrs[1])
        if elem in range(l, r+1):
            return 1
        elif l < elem:
            left = mid + 1
        elif r > elem:
            right = mid - 1
    return 0
```

Let's add a small pre-processing step to **merge** the overlapping steps in the binary search algorithm.

In [297]:
def merge_ranges(ranges: list[str]) -> list[tuple[int, int]]:
    intervals = [tuple(map(int, r.split('-'))) for r in ranges]
    intervals.sort(key=lambda x: x[0])
    merged = []
    for interval in intervals:
        # Add the new interval to the merged list if the second element in the last interval is less than the new one.
        if not merged or merged[-1][1] < interval[0] - 1:
            merged.append(interval)
        else:
            # Else, update the last element.
            merged[-1] = (merged[-1][0], max(merged[-1][1], interval[1]))
    return merged

ranges = [
    "3-5",
    "10-14",
    "16-20",
    "12-18",
]
print(merge_ranges(ranges))

[(3, 5), (10, 20)]


In [298]:
def bin_search(ranges: list[tuple[int, int]], elem: int):
    left = 0
    right = len(ranges) - 1
    while left <= right:
        mid = (left + right) // 2
        mrs = ranges[mid]
        # left and right margins of the range
        l, r = int(mrs[0]), int(mrs[1])
        if elem in range(l, r+1):
            return 1
        elif l < elem:
            left = mid + 1
        elif r > elem:
            right = mid - 1
    return 0

def search(ranges: list[str], elem: int):
    ranges = sorted(ranges, key=lambda x: int(x.split("-")[0]))
    merged = merge_ranges(ranges)
    return bin_search(ranges=merged, elem=elem)

def f(input: list[str]) -> int:
    split_idx = input.index("")
    ranges, ids = input[:split_idx], input[split_idx + 1:]

    count = 0
    for id in ids:
        count += search(ranges=ranges, elem=int(id))
    return count

with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    split_idx = input.index("")
    ranges, ids = input[:split_idx], input[split_idx + 1:]

    count = 0
    for id in ids:
        count += search(ranges=ranges, elem=int(id))

    print(f"Count = {count}")

Count = 623


Merging the ranges worked in binary search! Now, the algorithm is $O(n \times log(n))$.

---

The second part is actually the `merged_ranges` subroutine we wrote!

In [299]:
input = [
    "3-5",
    "10-14",
    "16-20",
    "12-18"
]

ranges = merge_ranges(input)

In [300]:
count = 0
for (l, r) in ranges:
    count += len(range(l, r+1))
count

14

In [301]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    split_idx = input.index("")

    ranges = merge_ranges(input[:split_idx])
    count = 0
    for (l, r) in ranges:
        count += len(range(l, r+1))

    print(f"Count = {count}")

Count = 353507173555373


`353507173555373` is in fact the correct answer!